dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/LoopUnrolling.java

```diff
@@ -97,15 +98,37 @@ private int unrollThreadAndUpdate(Thread t, int bound, int nextId){
 
     private void unrollThread(Thread thread, int defaultBound) {
         Event cur = thread.getEntry();
+        Map<Label, Integer> boundAnnotationMap = new HashMap<>();
+        LoopBound curBoundAnn = null;
 
         while (cur != null) {
             Event next = cur.getSuccessor();
-            if (cur instanceof CondJump && ((CondJump) cur).getLabel().getOId() < cur.getOId()) {
+            if (cur instanceof LoopBound && ((LoopBound) cur).getBound() > 0) {
+                if (curBoundAnn != null) {
+                    logger.warn("Found loop bound annotation that overwrites a previous, unused annotation.");
+                }
+                curBoundAnn = (LoopBound) cur;
+            } else if (curBoundAnn != null && cur instanceof Label) {
+                Label label = (Label)cur;
+                if (label.getJumpSet().stream().anyMatch(jump -> jump.getOId() > label.getOId())) {
+                    // The label denotes the beginning of a loop, and we found a LoopBound annotation before,
+                    // so we remember the bound and reset the annotation tracker.
+                    boundAnnotationMap.put(label, curBoundAnn.getBound());
+                    curBoundAnn = null;
+                }
```

Comment:
The else case corresponds to a `LoopBound` followed by a `Label` which is not really a jump. Can it be there there is another another Label between the bound and the actual start of the loop?

Reply:
The code matches a `LoopBound` annotation with the next `Label` that belongs to a loop. It doesn't matter how many non-looping `Label`s are inbetween.